[inductor] Fix TORCHINDUCTOR_FORCE_DISABLE_CACHES

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125863
* __->__ #129257

Summary: See https://github.com/pytorch/pytorch/issues/129159; this option wasn't doing its job for a few reasons. In this PR:
* Fix the with_fresh_cache_if_config() decorator
* Reset the "TORCHINDUCTOR_CACHE_DIR" & "TRITON_CACHE_DIR" env vars in sub-process to support them changing in the parent process

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @oulgen @jamesjwu @aorenste @anijain2305 @laithsakka

masnesral committed:
[inductor] Fix TORCHINDUCTOR_FORCE_DISABLE_CACHES

Summary: See https://github.com/pytorch/pytorch/issues/129159; this option wasn't doing its job for a few reasons. In this PR:
* Disable cache options in inductor's config.py if TORCHINDUCTOR_FORCE_DISABLE_CACHES is set.
* We can't directly disable the triton cache, so instead we use a new empty tmp location.
* We have to create that tmp location early, in the process pool warmup path and set the associated env var so that all subprocesses see that location.

[ghstack-poisoned]

oulgen commented on torch/_inductor/config.py:
```diff
@@ -27,23 +27,30 @@ def fx_graph_remote_cache_default():
 
+# Force disabled all inductor level caching -- This will override any other caching flag
+force_disable_caches = os.environ.get("TORCHINDUCTOR_FORCE_DISABLE_CACHES") == "1"
+_enable_caches = not force_disable_caches
```
why do we need this?

masnesral replied:
Only because it thought it was easier to read vs. a double negative:

`some_cache_enabled = _enable_caches and ...`
vs.
`some_cache_enabled = not force_disable_caches and ...`

oulgen commented on torch/_inductor/config.py:
```diff
@@ -27,23 +27,30 @@ def fx_graph_remote_cache_default():
+
 # use fx aot graph codegen cache
-fx_graph_cache = os.environ.get("TORCHINDUCTOR_FX_GRAPH_CACHE") == "1"
+fx_graph_cache = (
```
i dont think this is enough actually. This is decided at import time, so if you later set config.force_disable_caches = True, it will not have an impact on fx_graph_cache

oulgen replied:
You want to support

```
import torch

config.force_disable_caches = False

torch.compile(...)
```

use case

masnesral replied:
I actually kinda didn't want to support that, but ok, I guess I was being lazy. In that case, the way these settings are structured I don't know how to do it except update every point in code that checks for some cache being enabled. I wish I knew a better way.

masnesral replied:
Oh, I remembered (partly) why I was being lazy. After you import torch, the process pool has already started so the `config.force_disable_caches = False` is too late :/

oulgen replied:
> update every point in code that checks for some cache being enabled

This is kinda what i did for most of them, which you undid in this PR.

> After you import torch, the process pool has already started so the config.force_disable_caches = False is too late

is there no way to broadcast to the thread pool?

masnesral replied:
Yeah, I'd convinced myself that it was fine to expose this feature only via env var; I guess because it makes the problem much easier. It would also be much easier if Triton just allowed disabling the cache.

> is there no way to broadcast to the thread pool?

They're actually sub-processes. I think it would have to be a custom protocol. 

Still thinking...

oulgen replied:
> It would also be much easier if Triton just allowed disabling the cache.

We can add this, happy to write a PR to Triton for this. cc: @bertmaher @htyu

masnesral replied:
Actually I think I can work around it using the existing approach of creating temp dirs. It would be a bit cleaner if we get the ability to turn off the Triton cache in the future.

htyu replied:
Triton provides a way to bypass caching by always recompiling the kernel via `TRITON_ALWAYS_COMPILE=1`. Would that be helpful?

masnesral replied:
@htyu, aha, yes it would. Thanks!

masnesral committed:
Update on "[inductor] Fix TORCHINDUCTOR_FORCE_DISABLE_CACHES"


Summary: See https://github.com/pytorch/pytorch/issues/129159; this option wasn't doing its job for a few reasons. In this PR:
* Fix the with_fresh_cache_if_config() decorator
* Reset the "TORCHINDUCTOR_CACHE_DIR" & "TRITON_CACHE_DIR" env vars in sub-process to support them changing in the parent process

[ghstack-poisoned]

oulgen commented on torch/_inductor/async_compile.py:
```diff
@@ -172,11 +172,16 @@ def triton(self, kernel_name: str, source_code: str, device_str: str = "cuda"):
+            # We want to support changing these env vars after (and while) the
+            # process pool is running, so pass them to the subprocess to reset.
+            env_vars = ["TORCHINDUCTOR_CACHE_DIR", "TRITON_CACHE_DIR"]
+            extra_env = {v: os.environ.get(v, "") for v in env_vars if v in os.environ}
```
```suggestion
            extra_env = {v: os.environ[v] for v in env_vars if v in os.environ}
```
you already checked for membership, we shouldn't need the default anymore

masnesral replied:
It's only to satisfy the typechecker (which otherwise thinks get can return None)

oulgen replied:
If you [] access instead of .get, it cant return none since [] throws KeyError

oulgen commented on torch/_inductor/compile_fx.py:
```diff
@@ -420,13 +420,18 @@ def get_patched_config_dict(config_patches=None) -> Dict[str, Any]:
         return config.get_config_copy()
 
 
-@functools.wraps
```
what was wrong with this?

oulgen replied:
I dont really understand why you need to take arguments, instead of just function pointer here

oulgen replied:
also you probably want to keep wraps around so that stacktraces do not change

masnesral replied:
> what was wrong with this?

The previous implementation is checking `force_disable_caches` once when it evaluates the decoration, but then just returns the original function `compile_fx_inner`, which is executed directly thereafter. What we want is to return a function does the fresh_cache and then calls the decorated function.

For example, this test program:

```
disable = True

def decorator(f):
    if disable:
        print("Disabled")
        return f
    else:
        print("Enabled")
        return f

@decorator
def hello():
    print("hello world")

hello()
disable = False
hello()
```

Prints this output:
```
Disabled
hello world
hello world
```

But yeah, I accidentally removed the functools.wraps

oulgen commented on torch/_inductor/runtime/compile_tasks.py:
```diff
@@ -62,8 +62,11 @@ def _set_triton_ptxas_path() -> None:
     _set_triton_ptxas_path()
+
+    # Set any provided env vars
+    for k, v in extra_env.items():
+        os.environ[k] = v
+
```
```suggestion
    os.environ.update(extra_env)

```

masnesral committed:
comments on "[inductor] Fix TORCHINDUCTOR_FORCE_DISABLE_CACHES"


Summary: See https://github.com/pytorch/pytorch/issues/129159; this option wasn't doing its job for a few reasons. In this PR:
* Fix the with_fresh_cache_if_config() decorator
* Reset the "TORCHINDUCTOR_CACHE_DIR" & "TRITON_CACHE_DIR" env vars in sub-process to support them changing in the parent process

[ghstack-poisoned]

clee2000 commented:
@pytorchbot revert -m "one of the PRs in the stack seems to have broken test/distributed/_composable/test_replicate_with_compiler.py::ReplicateTest::test_bucketing_concat_op on distributed https://github.com/pytorch/pytorch/actions/runs/9653941844/job/26627760340 https://hud.pytorch.org/pytorch/pytorch/commit/4c1e4c5f307f9743014a08cf97d3fa8de7e1ce5f, not tested on this PR due to bad TD" -c weird

masnesral commented:
@clee2000 is it safe to just try again to merge? I'm not sure what "bad TD" means. But I'm fairly certain this change wouldn't affect the failing test and I can't repro a failure locally, so it seems miscategorized.

masnesral commented:
@pytorchbot merge
